### PR TITLE
txn: amend pessimistic transactions for modify column ddl

### LINKED ISSUE #18356: Support commit pessimistic transaction with modify/change column type concurrent DDLs
## Feature Request

**Is your feature request related to a problem? Please describe:**

Currently, the two-phase commit will fail if there is a concurrent DDL on related tables before commit.

**Describe the feature you'd like:**

Try to amend transaction memory buffer before committing for pessimistic transactions, if there are concurrent DDLs with drop index type.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

Subtask of https://github.com/pingcap/tidb/projects/50.

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What problem does this PR solve?

Issue Number: close #18356 
project: https://github.com/pingcap/tidb/projects/50

Problem Summary:

### What is changed and how it works?

How it Works:

For the modify/change DDL type, do check in the amend process for pessimistic transaction commit,  report error if there is column type changes other than increasing column field length.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

### Release note 
- Support commit pessimistic transactions with concurrent change column DDLs.

### CODE REVIEW

File: session/schema_amender.go [hunk 1]
```diff
@@ -39,7 +40,7 @@ import (
 
 const amendableType = nonMemAmendType | memBufAmendType
 const nonMemAmendType = (1 << model.ActionAddColumn) | (1 << model.ActionDropColumn) | (1 << model.ActionDropIndex)
-const memBufAmendType = uint64(1 << model.ActionAddIndex)
+const memBufAmendType = uint64(1<<model.ActionAddIndex) | (1 << model.ActionModifyColumn)
```
Comment (jackysp) [inline session/schema_amender.go:43]: It looks strange, but it's fine.

File: session/schema_amender.go [hunk 2]
```diff
@@ -128,6 +130,72 @@ func (a *amendCollector) keyHasAmendOp(key []byte) bool {
...
+		if colAtStart.Charset != colAtCommit.Charset || colAtStart.Collate != colAtCommit.Collate {
+			return errors.Trace(errors.Errorf("charset or collate is not matched for column=%v", colAtCommit.Name.String()))
+		}
+		err := ddl.CheckModifyTypeCompatible(&colAtStart.FieldType, &colAtCommit.FieldType)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	// TODO Default value change is not supported.
+	if !reflect.DeepEqual(colAtStart.DefaultValue, colAtCommit.DefaultValue) {
```
Comment (zimulala) [inline session/schema_amender.go:177]: Is it possible that the default value changes from oldDefaultValue to newDefaultValue and then to oldDefaultValue？

Comment (cfzjywxk) [reply session/schema_amender.go:177]: It's ok if they are the same at last, but I'm not sure if it'ok to check default value change in this way.

Comment (zimulala) [reply session/schema_amender.go:177]: Yes, even if `modify column c (new defaultValue)` -> `add index(c)` -> `modify column c (old defaultValue)`. Becuase these operations all in pessimistic transactions.

File: session/schema_amender.go [hunk 3]
```diff
@@ -128,6 +130,72 @@ func (a *amendCollector) keyHasAmendOp(key []byte) bool {
...
+			return errors.Trace(errors.Errorf("flag is not matched for column=%v, from=%v to=%v",
+				colAtCommit.Name.String(), colAtStart.FieldType.Flag, colAtCommit.FieldType.Flag))
+		}
+		if colAtStart.Charset != colAtCommit.Charset || colAtStart.Collate != colAtCommit.Collate {
+			return errors.Trace(errors.Errorf("charset or collate is not matched for column=%v", colAtCommit.Name.String()))
+		}
+		err := ddl.CheckModifyTypeCompatible(&colAtStart.FieldType, &colAtCommit.FieldType)
+		if err != nil {
+			return errors.Trace(err)
+		}
```
Comment (zimulala) [inline session/schema_amender.go:174]: It seems that we need to consider `auto rand bit`.

Comment (cfzjywxk) [reply session/schema_amender.go:174]: Will it be some problems if we commit using old `auto rand bit`? I'm not very sure.

Comment (zimulala) [reply session/schema_amender.go:174]: `auto-rand bit` operation is similar to `auto-increment id`. Refer to [auto-random](https://docs.pingcap.com/zh/tidb/stable/auto-random).

File: session/schema_amender.go (hunk 2 above)
Comment (zimulala) [reply session/schema_amender.go:177]: Could we add a test like `modify column c (new defaultValue)` -> `add index(c)` -> `modify column c (old defaultValue)`？
Add a test like the default value is changed to `now()`?

**Comment by cfzjywxk** [comment]
@tiancaiamao @coocood @lysu PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**